coding/python_style_guide.rst

```diff
@@ -12,648 +13,671 @@ To request changes to these standards, please file an :ref:`RFC <decision-making
+A mutable object MUST NOT be used as a keyword argument default
+---------------------------------------------------------------
 
-The Python language has evolved with time.
-Learn the new features of Python and use them where appropriate to make your code simpler and more readable.
-For example:
+Never use a mutable object as default value for a keyword argument in a function or method.
 
-- Use iterators, generators (classes that act like iterators) and generator expressions (expressions that act like iterators) to iterate over large data sets efficiently.
-  (New in Python 2.2, except generator expressions were added in 2.4 and generators were slightly enhanced in Python 2.5.)
+When used a mutable is used as a default keyword argument, the default *can* change from one call to another leading to unexpected behavior.
+This issue can be avoided by only using immutable types as default.
 
-- Use the ``with`` statement to simplify resource allocation.
-  (New in Python 2.5.)
-  For example to be sure a file will be closed when you are done with it: 
-  
-  .. code-block:: py
+For example, rather than provide a default empty list:
 
-     with open('/etc/passwd', 'r') as f:
-         for line in f:
-             pass
+.. code-block:: py
 
-The LSST environment currently supports Python 2.7.x.
-Do not use features that are not available in these versions of Python.
+   def proclist(alist=[]):
+       pass
 
-.. _style-guide-py-8-2:
+this function should create a new list in its internal scope:
 
-Python 2.5 improved Exception Handling SHOULD be used
------------------------------------------------------
+.. code-block:: py
 
-To catch all errors but let :py:exc:`~exceptions.SystemExit` and :py:exc:`~exceptions.KeyboardInterrupt` through, use:
+   def proclist(alist=None):
+       if alist is None:
+           alist = []
 
-.. code-block:: py
+.. _style-guide-py-context-managers:
 
-   except Exception, e:
-       pass
+Context managers (``with``) SHOULD be used for resource allocation
+------------------------------------------------------------------
 
-The exception hierarchy in Python 2.5 was improved, eliminating the need to use this: 
+Use the ``with`` statement to simplify resource allocation.
 
+For example to be sure a file will be closed when you are done with it: 
+  
 .. code-block:: py
 
-   except (SystemExit, KeyboardInterrupt):
-       raise
-       except Exception, e:
+   with open('/etc/passwd', 'r') as f:
+       for line in f:
            pass
 
-.. _style-guide-py-8-3:
+.. _style-guide-py-open:
 
-``raise ValueError('message')`` SHOULD be used instead of the deprecated form
------------------------------------------------------------------------------
+Use ``open`` instead of ``file``
+--------------------------------
 
-When raising an exception, use ``raise ValueError('message')`` instead of the older, deprecated form ``raise ValueError, 'message'``.
+This is preferred and ``file`` is gone in Python 3.
```

Comment:
I think I'd just drop the "This is preferred and" prefix because we not giving people a choice.